runsc/container/container_test.go

```diff
@@ -3178,3 +3187,156 @@ find $dir -type l -o -type f | sort | xargs cat | md5sum`), 0755); err != nil {
 		}
 	}
 }
+
+// createRootfsEROFS creates a rootfs directory and an EROFS rootfs image in
+// the directory dir.
+func createRootfsEROFS(dir string) (string, string, error) {
+	mkfs, err := exec.LookPath("mkfs.erofs")
+	if err != nil {
+		return "", "", fmt.Errorf("mkfs.erofs is not available: %v", err)
+	}
+
+	busybox, err := exec.LookPath("busybox")
+	if err != nil {
+		return "", "", fmt.Errorf("busybox is not available: %v", err)
+	}
+
+	// Create a rootfs directory with busybox in root.
+	rootfsDir := filepath.Join(dir, "rootfs")
+	for _, dir := range []string{"dev", "proc", "sys", "tmp"} {
+		if err := os.MkdirAll(filepath.Join(rootfsDir, dir), 0755); err != nil {
+			return "", "", fmt.Errorf("os.MkdirAll() failed: %v", err)
+		}
+	}
```

Comment:
Why are these needed? Does busybox need them? These will anyways be hidden by the sentry mounts at these locations.

Reply:
This is needed for now, because EROFS doesn't support creating synthetic directories yet. It's possible to workaround this by adding a writable overlay for rootfs (this is also what we did in [the EROFS checkpoint/restore test](https://github.com/google/gvisor/blob/cae689ed910df140ad3a6e78f7adeed308b57ee6/runsc/container/container_test.go#L3332-L3337)). But we can't do this in [the `TestRootfsEROFS` test](https://github.com/google/gvisor/blob/cae689ed910df140ad3a6e78f7adeed308b57ee6/runsc/container/container_test.go#L3246-L3256), because in that test, we want to check whether rootfs is EROFS via `/proc/self/mountinfo`. I will add some comments for this.

Reply:
Ah right.